### ISSUE TITLE
Incorrect urlencode of WebService target

### ISSUE THREAD

**Issue by Master-Inspire** [NLog/NLog#1629]
# Incorrect urlencode of WebService target

**Type**:Bug
**NLog version**: 4.3.8
**Platform**: .Net 4.6
**Current NLog config**

``` xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      autoReload="true"
      throwExceptions="false"
      internalLogLevel="Off" internalLogFile="c:\temp\nlog-internal.log">
    <targets>
        <target xsi:type='WebService'
                name='ws'
                url='http://localhost:3333/api/logme'
                protocol='HttpPost'
                encoding='UTF-8'   >
            <parameter name='param1' type='System.String' layout='${message}'/>
            <parameter name='param2' type='System.String' layout='${logger}'/>
        </target>
    </targets>
    <rules>
        <logger name='*' writeTo='ws'></logger>
    </rules>
</nlog>
```

**Api Controller**

``` csharp
public class LogMeController: ApiController {
    /// <summary>
    /// We need a complex type for modelbinding because
    /// of content-type: "application/x-www-form-urlencoded"
    /// </summary>
    public class ComplexType {
        public string Param1 { get; set; }
        public string Param2 { get; set; }
    }

    public void Post([FromBody] ComplexType complexType) {
        Console.WriteLine(complexType);
    }
}
```

Just use the [WebService target](https://github.com/NLog/NLog/wiki/WebService-target) as an example.
**Controller, write some special characters**

`````` csharp
public class SimpleController: Controller {
    private static readonly ILogger _logger = LogManager.GetCurrentClassLogger();

    public ActionResult Index() {
        _logger.Error("《∠∠⊙⌒∈∽》`````");
        return new ContentResult() {
            Content = "Simple"
        };
    }
}
``````

**And result**
![1](https://cloud.githubusercontent.com/assets/9047578/18355795/dc011dee-761c-11e6-905f-ac4eb35991de.png)
**NLog's source _src\NLog\Targets\WebServiceTarget.cs_ at line 375**

``` csharp
foreach (MethodCallParameter parameter in this.Parameters)
{
    sw.Write(separator);
    sw.Write(parameter.Name);
    sw.Write("=");
    sw.Write(UrlHelper.UrlEncode(Convert.ToString(parameterValues[i], CultureInfo.InvariantCulture), true));
    separator = "&";
    i++;
}
```

If I change the `UrlHelper.UrlEncode` method to `System.Net.WebUtility.UrlEncode` method, then i can get the `Param1` correctly decoded. So I think the `UrlHelper.UrlEncode`  may not correct. 

`UrlHelper.UrlEncode` and `System.Net.WebUtility.UrlEncode` is different.

`````` csharp
[TestFixture]
public class MyClass {
    [Test]
    public void TestMethod1() {
        var encode1 = UrlHelper.UrlEncode("《∠∠⊙⌒∈∽》`````", true);
        Console.WriteLine(encode1);
        // %u300a%u2220%u2220%u2299%u2312%u2208%u223d%u300b%60%60%60%60%60
        var encode2 = System.Net.WebUtility.UrlEncode("《∠∠⊙⌒∈∽》`````");
        Console.WriteLine(encode2);
        // %E3%80%8A%E2%88%A0%E2%88%A0%E2%8A%99%E2%8C%92%E2%88%88%E2%88%BD%E3%80%8B%60%60%60%60%60
    }
}
``````

`System.Net.WebUtility.UrlDecode` can not decode the string from `UrlHelper.UrlEncode`.

**Comment by 304NotModified** [comment]
Not sure of this is a bug, as there are indeed multiple ways to encode URLs. But I guess it would be nice to configure how it will encoded right?

**Comment by Master-Inspire** [comment]
@304NotModified I think it is a bug. For now my workaround is write a custom target and post as json instead .

Special characters means Chinese (eg. _中文_)or Japanese(eg. _ほとすぬに_) characters. 

Javascript `encodeURIComponent` function has the same output with `System.Net.WebUtility.UrlEncode`. 

I am wondering why use a `UrlHelper.UrlEncode` method and not `System.Net.WebUtility.UrlDecode`?

**Comment by 304NotModified** [comment]
> I am wondering why use a UrlHelper.UrlEncode method and not System.Net.WebUtility.UrlDecode?

Because of dependencies which are not available on every platform. 

I still think this is not a bug, for example:

![image](https://cloud.githubusercontent.com/assets/5808377/18394712/89d81604-76bb-11e6-81f7-37584378fd58.png)

![image](https://cloud.githubusercontent.com/assets/5808377/18394716/8fc0afea-76bb-11e6-8387-b80193dce2fe.png)


Not saying we should not change things! 

This is interesting though (not sure if this applies to our implementation)

> There exists a non-standard encoding for Unicode characters: %uxxxx, where xxxx is a UTF-16 code unit represented as four hexadecimal digits. This behavior is not specified by any RFC and has been rejected by the W3C

https://en.wikipedia.org/wiki/Percent-encoding

**Comment by 304NotModified** [comment]
Related link: http://stackoverflow.com/questions/21859233/url-encode-ascii-utf16-characters

@Master-Inspire Have you tried with the encoding utf16 instead of utf8?

**Comment by Master-Inspire** [comment]
`System.Web.HttpUtility.UrlEncode` have the same output with `System.Net.WebUtility.UrlEncode` (ignore case).

`System.Web.HttpUtility.UrlDecode` can decode NLog's `UrlHelper.UrlEncode` correctly, but `System.Net.WebUtility.UrlDecode` can't.

```csharp
var encode1 = UrlHelper.UrlEncode("《∠∠⊙⌒∈∽》`````", true);
Console.WriteLine(encode1);
// %u300a%u2220%u2220%u2299%u2312%u2208%u223d%u300b%60%60%60%60%60

Console.WriteLine(System.Web.HttpUtility.UrlDecode(encode1));
// 《∠∠⊙⌒∈∽》`````

Console.WriteLine(System.Net.WebUtility.UrlDecode(encode1));
// %u300a%u2220%u2220%u2299%u2312%u2208%u223d%u300b`````

var encode2 = System.Net.WebUtility.UrlEncode("《∠∠⊙⌒∈∽》`````");
Console.WriteLine(encode2);
// %E3%80%8A%E2%88%A0%E2%88%A0%E2%8A%99%E2%8C%92%E2%88%88%E2%88%BD%E3%80%8B%60%60%60%60%60

var encode3 = System.Web.HttpUtility.UrlEncode("《∠∠⊙⌒∈∽》`````");
Console.WriteLine(encode3);
// %e3%80%8a%e2%88%a0%e2%88%a0%e2%8a%99%e2%8c%92%e2%88%88%e2%88%bd%e3%80%8b%60%60%60%60%60

```
I have to **manually** decode `Param1` use `System.Web.HttpUtility.UrlDecode`. :disappointed: **That's not good.**
![1](https://cloud.githubusercontent.com/assets/9047578/18575257/1a4b1258-7c08-11e6-9e14-4473b2e62c7b.png)

@304NotModified Maybe you should look at **ServiceStack.Text** `UrlEncode` implementation:
https://github.com/ServiceStack/ServiceStack.Text/blob/master/src/ServiceStack.Text/StringExtensions.cs#L127

**Comment by 304NotModified** [comment]
Any idea if there are any RFCs for URL encode? It difficult that there are multiple implementations and unclear specs

**Comment by 304NotModified** [comment]
RFCs:

> By default, the string is escaped according to RFC 2396. If International Resource Identifiers (IRIs) or Internationalized Domain Name (IDN) parsing is enabled, the string is escaped according to RFC 3986 and RFC 3987

https://msdn.microsoft.com/en-us/library/system.uri.escapedatastring(v=vs.110).aspx

**Comment by snakefoot** [comment]
Maybe resolved with  #1735 ?

**Comment by 304NotModified** [comment]
Yes it is!
